### PR TITLE
Fix sync job which fell over when synchronisation of all the subjects was attempted

### PR DESCRIPTION
**Pull request by laurieboyes**
The key problems this PR solves are:
- Used directly to apply the concept map sequentially, preventing loads of relationship deletion queries being executed simultaneously
- Stopped attempts to delete `Concept` node when it still had loads of `Event` nodes related to it
- Stages `followed` relationship deletions (and new `Event` node deletions) out so that they are attempted in batches of 10, which stops these queries timing out and piling up.  

Note: Only concepts which current have users following them are affected by this job, so some lonely subject concepts will remain

It also, renames the `AllNodeQueries` to `QueryBatch`, which I think is much better (even though the sync job didn't end up using this in the end)

### CODE REVIEW

File: jobs/_sync.js [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+			let queryParams = [];
+			let batches = Math.ceil(res[0].count / 1000);
+
+			while (batches--) {
+				queryParams.push({start: (batches) * 1000});
+			}
+			return directly(5, queryParams.map(q => () => {
+				worker.metrics.count('concept_batch_query');
+				return db.cypher({
+					query: `MATCH (concept:Concept) RETURN concept SKIP {start} LIMIT 1000`,
```
Comment (phamann) [inline jobs/_sync.js]: Why are you SKIP'ing here and not using flagging?

File: jobs/_sync.js [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+function getConceptMap (concepts) {
+
+	// put concepts into batches of 10
+	const conceptBatches = [];
+	while (concepts.length) {
+		conceptBatches.push(concepts.splice(0, 10));
+	}
+
+	const conceptMappings = new Map();
```
Comment (phamann) [inline jobs/_sync.js]: :heart_eyes_cat:

File: jobs/_sync.js [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+}
+
+
+function deleteConceptEventsCarefully (conceptId) {
+
+	return db.cypher({
+		query: `
+			MATCH (c:Concept {uuid: {uuid}})
+			MATCH (event)-[r:subject]->(c)
+			WITH c, event LIMIT 10
```
Comment (phamann) [inline jobs/_sync.js]: 10 seems quite conservative?

File: jobs/_sync.js [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+			}
+		});
+}
+
+function removeConceptRelationshipsCarefully (conceptId) {
+
+	return db.cypher({
+		query: `
+			MATCH (c:Concept {uuid: {uuid}})
+			MATCH (actor)-[r]->(c)
```
Comment (phamann) [inline jobs/_sync.js]: Won't `(actor)` match all nodes? Shouldn't you reverse query to scope to sub-graph i.e. `(c)<-[r]-(actor)`

File: jobs/_sync.js (hunk 1 above)
Comment (laurieboyes) [inline jobs/_sync.js]: Hm. Laziness? Although the SKIP jumped out at us originally, it doesn't seem to have ever caused a problem, so I haven't altered this part of the job at all. Reckon it's worth it?

File: jobs/_sync.js [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+				// copy all followed relationships pointing at the old concept to the new one
+				// then delete the old concept and relationships
+				return () => db.cypher({
+					query: `
+							MERGE (new:Concept {uuid: {newUuid}})
+							SET new += {newProps}
+							WITH new
+							MATCH (old:Concept {uuid: {oldUuid}})<-[oldRel:followed]-(user)
+							WITH collect(user) as users, new, old, oldRel
+							FOREACH (u in users | MERGE u-[newRel:followed]->new SET newRel += oldRel )
```
Comment (phamann) [inline jobs/_sync.js]: unwrapped nodes in queries are being deprecated. Needs to read `(u)-[newRel:followed]->(new)`

**Comment by phamann** [comment]
:+1: Looks great!

File: jobs/_sync.js (hunk 3 above)
Comment (laurieboyes) [inline jobs/_sync.js]: Aye. I bumped it up to 50 and found it was still occasionally timing out and causing problems. Slow and steady wins the race.

File: jobs/_sync.js (hunk 5 above)
Comment (laurieboyes) [inline jobs/_sync.js]: Yes! Good spot

File: jobs/_sync.js (hunk 4 above)
Comment (laurieboyes) [inline jobs/_sync.js]: From my experiments it seems that as `c` is already in scope, the scan is contained to the sub-graph regardless of the order of the nodes in the relationship

**Pull request merged**